sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,311 @@
+
+// Validate PLIC priority
+static bool plic_validate_priority(uint32_t priority) {
+  return (priority <= RV_PLIC_MAX_PRIORITY);
+}
+
+// Check PLIC state for a claimed but not yet processed IRQ
+static bool plic_irq_claimed(dif_plic_t *plic) {
+  return (plic->claimed_irq.peripheral != kDifPlicPeripheralNone &&
+          plic->claimed_irq.source != kDifPlicNoIrq);
+}
+
+// Clear PLIC state claimed IRQ data
+static void plic_irq_claimed_clear_data(dif_plic_t *plic) {
+  plic->claimed_irq.peripheral = kDifPlicPeripheralNone;
+  plic->claimed_irq.source = kDifPlicNoIrq;
+}
+
+// Get IE, IP or LE register offset (IE00, IE01, ...) from IRQ source ID.
+static ptrdiff_t plic_offset_from_reg0(dif_plic_irq_src_id_t irq) {
+  uint8_t register_index = (uint32_t)irq / PLIC_ID_TO_INDEX_MODULO;
+  return register_index * sizeof(uint32_t);
+}
+
+// Get PRIO register offset (PRIO0, PRIO1, ...) from IRQ source ID.
+static ptrdiff_t plic_priority_offset_from_reg0(dif_plic_irq_src_id_t irq) {
+  uint8_t register_index = (uint32_t)irq;
+  return register_index * sizeof(uint32_t);
+}
+
+// Get IE, IP, LE register bit index from IRQ source ID. IRQ IDs start from 1,
+// however IRQ bit index inside the registers start from 0. This means that in
+// order to get IRQ bit index in the register, 1 has to be substracted from the
+// obtained remainder.
+static uint8_t plic_bit_index_from_irq_id(dif_plic_irq_src_id_t irq) {
+  return ((uint32_t)irq % PLIC_ID_TO_INDEX_MODULO) - 1;
+}
+
+// Get the target that will be interrupted by the IRQ source
+static dif_plic_target_t plic_irq_target_from_src_id(
+    dif_plic_irq_src_id_t irq) {
+  if (irq >= PLIC_IBEX_SOURCE_FIRST_IRQ && irq <= PLIC_IBEX_SOURCE_LAST_IRQ) {
+    return kDifPlicIbex0;
+  }
+
+  return kDifPlicTargetNone;
+}
+
+bool dif_plic_init(uintptr_t base_addr, dif_plic_target_t target,
+                   dif_plic_t *plic) {
+  if (plic == NULL || base_addr == 0) {
+    return false;
+  }
+
+  // Validate the requested target, and initialise the target specific registers
+  reg32_t addr = reg32_from_addr(base_addr);
+  switch (target) {
+    case kDifPlicIbex0:
+      reg32_write(addr, RV_PLIC_IE00, 0u);
+      reg32_write(addr, RV_PLIC_IE01, 0u);
+      reg32_write(addr, RV_PLIC_THRESHOLD0, 0u);
+      break;
+    default:
+      return false;
+  };
+
+  plic->target = target;
+  plic->base_addr = addr;
+  plic_irq_claimed_clear_data(plic);
+
+  return true;
+}
+
+bool dif_plic_irq_ie_control(const dif_plic_t *plic, dif_plic_irq_src_id_t irq,
+                             dif_plic_on_off_t enable) {
+  if (plic == NULL || !plic_irq_source_validate(irq) ||
+      !plic_on_off_valid(enable)) {
+    return false;
+  }
+
+  // Requested interrupt source is not connected to the PLIC target
+  if (plic_irq_target_from_src_id(irq) != plic->target) {
+    return false;
+  }
+
+  // Get the target and IRQ source specific Interrupt Enable register
+  ptrdiff_t offset = plic_offset_from_reg0(irq);
+  ptrdiff_t ie_offset;
```

Comment:
Expand `ie` to `interrupt_enable`.

Please avoid other abbreviations, like `le`, `ip`, and `prio`.